File: Cpp/Linked-List-Cycle-II.cpp
/**
 * Definition of ListNode
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(int val) {
        this->val = val;
        this->next = NULL;
    }
};

class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The node where the cycle begins.
     *           if there is no cycle, return null
     */
    ListNode *detectCycle(ListNode *head) {
        // write your code here
        if (head == NULL || head ->next == NULL) return NULL;
        ListNode *node1 = head, *node2 = head;
        node1 = node1->next;
        node2 = (node2->next)->next;
        if (node2 == NULL) return NULL;
        int cnt = 1;
        while (node1 != node2) {
            node1 = node1->next;
            node2 = node2->next;
            if (node2 == NULL) return NULL;
            node2 = node2->next;
            if (node2 == NULL) return NULL;
            cnt ++;
        }
        ListNode *node = head;
        for (int i = 0; i < cnt; i ++) {
            node = node->next;
        }
        node1 = head;
        while (node != node1) {
            node = node->next;
            node1 = node1->next;
        }
        return node;
    }
};

File: Java/Linked-List-Cycle-II.java
/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */ 
public class Solution {
    /**
     * @param head: The first node of linked list.
     * @return: The node where the cycle begins. 
     *           if there is no cycle, return null
     */
    public ListNode detectCycle(ListNode head) {
        // write your code here
        if (head == null || head.next == null) return null;
        ListNode slow = head;
        ListNode fast = head.next;
        int len = 1;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            len ++;
            if (slow == fast) break;
        }
        if (fast == null || fast.next == null) return null;
        ListNode cur = head;
        slow = head;
        for (int i = 0; i < len; i ++) {
            cur = cur.next;
        }
        while (slow != cur) {
            slow = slow.next;
            cur = cur.next;
        }
        return cur;
    }
}

